Doc/howto/concurrency.rst

```diff
@@ -0,0 +1,1846 @@
+of each interpreter is effectively isolated from the others.
+
+That isolation includes things like :data:`sys.modules`.  By default,
+interpreters mostly don't share any data (including objects) at all.
+Anything that gets shared is done on a strictly opt-in basis.  That
+means programmers wouldn't need to worry about possible races with
+*any* data in the program.  They would only need to worry about data
+that was explicitly shared.
+
+Interpreters themselves are not specific to any thread, but instead
+each physical thread has (at most) one interpreter active at any given
+moment.  Each interpreter can be associated in this way with any number
+of threads.  Since each interpreter is isolated from the others,
+any thread using one interpreter is thus isolated from threads
+using any other interpreter.
+
+Using multiple interpreters is fairly straight-forward:
+
+1. create a new interpreter
+2. switch the current thread to use that interpreter
+3. call :func:`exec`, but targeting the new interpreter
+4. switch back
+
+Note that no threads were involved; running in an
+interpreter happens relative to the current thread.  New threads
+aren't implicitly involved.
+
+Multi-processing and distributed computing provide similar isolation,
+though with some tradeoffs.
+
+.. _python-stdlib-interpreters:
+
+A stdlib module for using multiple interpreters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+While use of multiple interpreters has been part of Python's C-API
+for decades, the feature hasn't been exposed to Python code through
+the stdlib.  :pep:`734` proposes changing that by adding a new
+:mod:`!interpreters` module.
+
+In the meantime, an implementation of that PEP is available for
+Python 3.13+ on PyPI: :pypi:`interpreters-pep-734`.
+
+.. _python-interpreters-overhead:
+
+Improving performance for multiple interpreters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The long-running effort to improve on Python's implementation of multiple
+interpreters focused on isolation and stability; very little done
+to improve performance.  This has the most impact on:
+
+* how much memory each interpreter uses
+  (i.e. how many can run at the same time)
+* how long it takes to create a new interpreter
+
+It also impacts how efficiently data/objects can be passed between
+interpreters, and how effectively objects can be shared.
+
+As the work on isolation wraps up, improvements will shift to focus
+on performance and memory usage.  Thus, the overhead of
+using multiple interpreters will drastically decrease over time.
+
+Shared resources
+^^^^^^^^^^^^^^^^
+
+Aside from memory, all physical threads in a process share the
+following resources:
+
+* command line arguments ("argv")
+* env vars
+* current working directory
+* signals, IPC, etc.
+* open I/O resources (file descriptors, sockets, etc.)
+
+When relevant, these must be managed in a thread-safe way.
+
+Tracing execution
+^^^^^^^^^^^^^^^^^
+
+TBD
+
+.. TODO finish
```

Comment:
I always mention "determinism" as one of my favourite assets of async in comparison with threading. The following sentences are not to explain this to *you*, but what I use to explain to people learning about this, maybe some of this can be incorporated here or in other part.

So...

In contrast with threading, where "something" (Python, the OS, the CPU) can take you out from the processor, in asyncio as it's a collaborative model you always have to yield it. Of course, you need to be sure to not be too greedy about it, but also means that the code you read sequentially, it happens sequentially. 

In other words, if you have a logger call in line 5 and other logger call in line 27, when you're trying to debug what happened to a run of your program, and you check a log file, as you see the corresponding log line to logger call in code line 5, and the you see the corresponding log line to logger call in code line 27, you *are sure* that nothing got in the middle. This way, you can think of that part of the code sequentially, nothing changed any state, no objects values where modified, nobody messed with your system between those lines. And that is a HUGE difference with threading model, where you read those lines of code and you know that unless somehing is wrapped in a lock, a million things could have happened between line N  and line N+1.

Reply:
Yeah, I got started on this section but didn't feel like my thoughts were enough organized to be clear yet. 😄  I'll see about incorporating your explanation.